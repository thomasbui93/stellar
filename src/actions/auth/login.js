import { FAILED_AUTHENTICATION, AUTHENTICATED, DATA_IS_LOADING } from './constants'
import {post} from '../../utils/request'

export const requestAuthentication = (username, password) => {
  return function (dispatch) {
    dispatch(dataIsLoading(true))
    return post('authenticate', {username: username, password: password})
      .then(response => {
        dispatch(authenticateSuccessfully(response))
      })
      .catch(() => {
        dispatch(authenticateFailed())
      })
      .then(() => {
        dispatch(dataIsLoading(false))
      })
  }
}

export function authenticateSuccessfully (response) {
  return {type: AUTHENTICATED, token: response.token}
}

export function authenticateFailed () {
  return {type: FAILED_AUTHENTICATION}
}

export function dataIsLoading (bool) {
  return {
    type: DATA_IS_LOADING,
    isLoading: bool
  }
}
