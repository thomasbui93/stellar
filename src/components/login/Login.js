import React, {Component} from 'react'
import PropTypes from 'prop-types'

export class Login extends Component {
  static propTypes = {
    onSubmit: PropTypes.func,
    isLoading: PropTypes.bool
  };

  constructor (props) {
    super(props)
    this.submit = this.submit.bind(this)
  }

  submit (event) {
    event.preventDefault()
    this.props.onSubmit(this.refs.username.value, this.refs.password.value)
  }

  render () {
    return <form className='login-form' onSubmit={this.submit}>
      <div className='field'>
        <label className='label'>Email</label>
        <div className='control'>
          <input className='input' type='text' placeholder='Username' ref='username' />
        </div>
      </div>
      <div className='field'>
        <label className='label'>Password</label>
        <div className='control'>
          <input className='input' type='password' ref='password' />
        </div>
      </div>
      <div className='control'>
        <button className='button is-primary' type='submit' disabled={this.props.isLoading}>Submit</button>
      </div>
    </form>
  }
}
