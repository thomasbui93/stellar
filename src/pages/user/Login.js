import React from 'react'
import { connect } from 'react-redux'
import { requestAuthentication } from '../../actions/user/auth';
import { withRouter } from 'react-router'
import Notice from '../../components/general/Notice'
import { saveToken } from '../../utils/auth';

export class LoginPage extends React.Component {
  constructor(props) {
    super(props)
    this.requestAuthentication = this.requestAuthentication.bind(this)
  }

  requestAuthentication(){
    return this.props.requestAuthentication(this.refs.username.value, this.refs.password.value)
  }
  
  componentWillReceiveProps(nextProps) {
    if(nextProps.token) {
      this.props.redirectToMainPage(nextProps.token)
      this.props.history.push('/categories')
    }
  }

  render() {
    return (
      <div>
        <h1> Log In </h1>
        { this.props.error ? <Notice type='error' message={this.props.error}/> : ''}
        <div className="field">
          <div className="control">
            <label>User Name</label>
          </div>
          <div className="input">
            <input type="text" ref="username"/>
          </div>
        </div> 
        <div className="field">
          <div className="control">
            <label>Password</label>
          </div>
          <div className="input">
            <input type="password" ref="password"/>
          </div>
        </div>
        <div className="action">
          <button onClick={this.requestAuthentication}> { this.props.isLoading ? 'Logging In' : 'Login' } </button>
        </div>        
      </div>
    )
  }
}

const mapStateToProps = ({ userAuth }) => {
  return {
    isLoading: userAuth.isLoading,
    token: userAuth.token,
    error: userAuth.error
  }
}

const mapDispatchToProps = (dispatch, ownProps) => {
  return {
    requestAuthentication: (username, password) => {
      dispatch(requestAuthentication(username, password))
    },
    redirectToMainPage: (token) => {
      saveToken(token)
    }
  }
}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(LoginPage))