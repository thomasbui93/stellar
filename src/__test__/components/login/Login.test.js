import React from 'react';
import {Login} from "../../../components/login/Login";
import renderer from 'react-test-renderer';
import {shallow, mount} from 'enzyme';
import {expect as expectChai} from 'chai';
import { spy } from 'sinon'

describe('Login Component', () => {
  it('render correctly', () =>{
    const component = renderer.create(<Login/>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render required elements: input, button', ()=> {
    const wrapper = shallow(<Login />);
    expectChai(wrapper.find('button')).to.have.length(1);
    expectChai(wrapper.find('input[type="text"]')).to.have.length(1);
    expectChai(wrapper.find('input[type="password"]')).to.have.length(1);
  });

  it('should disable submit button while isLoading is set to true', ()=>{
    const wrapper = shallow(<Login isLoading={true} />);
    expectChai(wrapper.find('button').prop('disabled')).equal(true);
  });

  it('should call onSubmit if button is click', () => {
    const onButtonClick = spy();
    const wrapper = mount(<Login isLoading={false} onSubmit={onButtonClick}/>);
    wrapper.find('form').simulate('submit');
    expectChai(onButtonClick.calledOnce).to.equal(true);
  });

});