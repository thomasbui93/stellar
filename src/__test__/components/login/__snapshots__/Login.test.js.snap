// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component render correctly 1`] = `
<form
  className="login-form"
  onSubmit={[Function]}
>
  <div
    className="field"
  >
    <label
      className="label"
    >
      Email
    </label>
    <div
      className="control"
    >
      <input
        className="input"
        placeholder="Username"
        type="text"
      />
    </div>
  </div>
  <div
    className="field"
  >
    <label
      className="label"
    >
      Password
    </label>
    <div
      className="control"
    >
      <input
        className="input"
        type="password"
      />
    </div>
  </div>
  <div
    className="control"
  >
    <button
      className="button is-primary"
      disabled={undefined}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;
